File: DevExpress VCL/ExpressPivotGrid/Demos/Delphi/Common/DemoBasicMain.pas
unit DemoBasicMain;

{$I ..\cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxLookAndFeels, StdCtrls, Menus, cxCustomPivotGrid, cxExportPivotGridLink;

type
  TfrmDemoBaisicMain = class(TForm)
    mmMain: TMainMenu;
    miFile: TMenuItem;
    miExport: TMenuItem;
    miExportToExcel: TMenuItem;
    miExportToText: TMenuItem;
    miExportToHtml: TMenuItem;
    miExportToXml: TMenuItem;
    Separator1: TMenuItem;
    miExit: TMenuItem;
    miOptions: TMenuItem;
    miLookFeel: TMenuItem;
    Kind1: TMenuItem;
    miFlat: TMenuItem;
    miStandard: TMenuItem;
    miUltraFlat: TMenuItem;
    miOffice11: TMenuItem;
    miNativeStyle: TMenuItem;
    miAbout: TMenuItem;
    SaveDialog: TSaveDialog;
    lbDescrip: TLabel;
    lfController: TcxLookAndFeelController;
    miTotalsPosition: TMenuItem;
    miTotalsLocationFar: TMenuItem;
    miTotalsLocationNear: TMenuItem;
    miShowTotalsForSingleValues: TMenuItem;
    miShowRowTotals: TMenuItem;
    miShowColumnTotals: TMenuItem;
    miShowColumnGrandTotals: TMenuItem;
    miShowRowGrandTotals: TMenuItem;
    miTotalsVisibility: TMenuItem;
    miShowGrandTotalsForSingleValues: TMenuItem;
    N2: TMenuItem;
    miElementsVisibility: TMenuItem;
    miShowFilterFields: TMenuItem;
    miShowColumnFields: TMenuItem;
    miShowDataFields: TMenuItem;
    miShowRowFields: TMenuItem;
    miShowFilterSeparator: TMenuItem;
    N1: TMenuItem;
    procedure miExportToClick(Sender: TObject);
    procedure miTotalsLocationClick(Sender: TObject);
    procedure miExitClick(Sender: TObject);
    procedure miTotalsVisibilityClick(Sender: TObject);
    procedure miElementsVisibilityClick(Sender: TObject);
    procedure miLookAndFeelClick(Sender: TObject);
    procedure miAboutClick(Sender: TObject);
  private
    { Private declarations }
  protected
    function GetPivotGrid: TcxCustomPivotGrid; virtual;
    procedure SyncElementsVisibilityWithMenu;
    procedure SyncTotalVisibilityWithMenu;
    procedure SyncMenuWithElementsVisibility;
    procedure SyncMenuWithTotalVisibility;
  public
    constructor Create(AOwner: TComponent); override;

    property PivotGrid: TcxCustomPivotGrid read GetPivotGrid;
  end;

implementation

uses
  AboutDemoForm;

{$R *.dfm}

constructor TfrmDemoBaisicMain.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
  Application.Title := Caption;
  PivotGrid.ApplyBestFit;
  SyncMenuWithTotalVisibility;
  SyncMenuWithElementsVisibility;
end;

function TfrmDemoBaisicMain.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := nil;
end;

procedure TfrmDemoBaisicMain.SyncElementsVisibilityWithMenu;
begin
  with PivotGrid.OptionsView do
  begin
    ColumnFields := miShowColumnFields.Checked;
    DataFields := miShowDataFields.Checked;
    FilterFields := miShowFilterFields.Checked;
    FilterSeparator := miShowFilterSeparator.Checked;
    RowFields := miShowRowFields.Checked;
  end;
end;

procedure TfrmDemoBaisicMain.SyncTotalVisibilityWithMenu;
begin
  with PivotGrid.OptionsView do
  begin
    ColumnTotals := miShowColumnTotals.Checked;
    RowTotals := miShowRowTotals.Checked;
    ColumnGrandTotals := miShowColumnGrandTotals.Checked;
    RowGrandTotals := miShowRowGrandTotals.Checked;
    TotalsForSingleValues := miShowTotalsForSingleValues.Checked;
    GrandTotalsForSingleValues := miShowGrandTotalsForSingleValues.Checked;
  end;
end;

procedure TfrmDemoBaisicMain.SyncMenuWithElementsVisibility;
begin
  with PivotGrid.OptionsView do
  begin
    miShowColumnFields.Checked := ColumnFields;
    miShowDataFields.Checked := DataFields;
    miShowFilterFields.Checked := FilterFields;
    miShowFilterSeparator.Checked := FilterSeparator;
    miShowRowFields.Checked := RowFields;
  end;
end;

procedure TfrmDemoBaisicMain.SyncMenuWithTotalVisibility;
begin
  with PivotGrid.OptionsView do
  begin
    miShowColumnTotals.Checked := ColumnTotals;
    miShowRowTotals.Checked := RowTotals;
    miShowColumnGrandTotals.Checked := ColumnGrandTotals;
    miShowRowGrandTotals.Checked := RowGrandTotals;
    miShowTotalsForSingleValues.Checked := TotalsForSingleValues;
    miShowGrandTotalsForSingleValues.Checked := GrandTotalsForSingleValues;
  end;
end;

procedure TfrmDemoBaisicMain.miExportToClick(Sender: TObject);
begin
  if SaveDialog.Execute then
  begin
    case TMenuItem(Sender).Tag of
      1:
        cxExportPivotGridToExcel(SaveDialog.FileName, PivotGrid);
      2:
        cxExportPivotGridToText(SaveDialog.FileName, PivotGrid);
      3:
        cxExportPivotGridToHTML(SaveDialog.FileName, PivotGrid);
      4:
        cxExportPivotGridToXML(SaveDialog.FileName, PivotGrid);
    end;
  end;
end;

procedure TfrmDemoBaisicMain.miTotalsLocationClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := True; 
  PivotGrid.OptionsView.TotalsLocation := TcxPivotGridTotalsLocation(TMenuItem(Sender).Tag);
  miTotalsLocationFar.Checked := PivotGrid.OptionsView.TotalsLocation = tlFar;
  miTotalsLocationNear.Checked := PivotGrid.OptionsView.TotalsLocation = tlNear;
end;

procedure TfrmDemoBaisicMain.miExitClick(Sender: TObject);
begin
  Application.Terminate;
end;

procedure TfrmDemoBaisicMain.miTotalsVisibilityClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  SyncTotalVisibilityWithMenu;
end;

procedure TfrmDemoBaisicMain.miElementsVisibilityClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  SyncElementsVisibilityWithMenu;
end;

procedure TfrmDemoBaisicMain.miLookAndFeelClick(Sender: TObject);
begin
  if TMenuItem(Sender).Tag > 3 then
    lfController.NativeStyle := not lfController.NativeStyle
  else
  begin
    TMenuItem(Sender).Checked := True;
    lfController.NativeStyle := False; 
    lfController.Kind := TcxLookAndFeelKind(TMenuItem(Sender).Tag);
  end;
  miNativeStyle.Checked := lfController.NativeStyle;
end;

procedure TfrmDemoBaisicMain.miAboutClick(Sender: TObject);
begin
  ShowAboutDemoForm;
end;

end.

File: DevExpress VCL/ExpressPivotGrid/Demos/CBuilder/Common/DemoBasicMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "DemoBasicMain.h"
#include "cxExportPivotGridLink.hpp"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxLookAndFeels"
#pragma resource "*.dfm"

//---------------------------------------------------------------------------
__fastcall TfrmDemoBaisicMain::TfrmDemoBaisicMain(TComponent* Owner)
  : TForm(Owner)
{
  Application->Title = Caption;
  SyncMenuWithTotalVisibility();
  SyncMenuWithElementsVisibility();
}

TcxCustomPivotGrid* __fastcall TfrmDemoBaisicMain::PivotGrid()
{
  return NULL;
}

void __fastcall TfrmDemoBaisicMain::miExportToClick(TObject *Sender)
{
  if (SaveDialog->Execute()) {
    switch (((TMenuItem*)Sender)->Tag){
      case 1:
        cxExportPivotGridToExcel(SaveDialog->FileName, PivotGrid(), true, true, "xls"); break;
      case 2:
        cxExportPivotGridToText(SaveDialog->FileName, PivotGrid(), true, "", "", "", "txt"); break;
      case 3:
        cxExportPivotGridToHTML(SaveDialog->FileName, PivotGrid(), true, "html"); break;
      case 4:
        cxExportPivotGridToXML(SaveDialog->FileName, PivotGrid(), true, "xml"); break;
    }
  }
}

void __fastcall TfrmDemoBaisicMain::miTotalsLocationClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = true;
  PivotGrid()->OptionsView->TotalsLocation =
    (TcxPivotGridTotalsLocation)((TMenuItem*)Sender)->Tag;
  miTotalsLocationFar->Checked = (PivotGrid()->OptionsView->TotalsLocation == tlFar);
  miTotalsLocationNear->Checked = (PivotGrid()->OptionsView->TotalsLocation == tlNear);
}

void __fastcall TfrmDemoBaisicMain::miExitClick(TObject *Sender)
{
  Application->Terminate();
}

void __fastcall TfrmDemoBaisicMain::miTotalsVisibilityClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  SyncTotalVisibilityWithMenu();
}

void __fastcall TfrmDemoBaisicMain::miElementsVisibilityClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  SyncElementsVisibilityWithMenu();
}

void __fastcall TfrmDemoBaisicMain::miLookAndFeelClick(TObject *Sender)
{
  if (((TMenuItem*)Sender)->Tag > 3) {
    lfController->NativeStyle = !lfController->NativeStyle;
  }
  else {
    ((TMenuItem*)Sender)->Checked = true;
    lfController->NativeStyle = false;
    lfController->Kind = (TcxLookAndFeelKind)((TMenuItem*)Sender)->Tag;
  }
  miNativeStyle->Checked = lfController->NativeStyle;
}

void __fastcall TfrmDemoBaisicMain::miAboutClick(TObject *Sender)
{
  ShowAboutDemoForm();
}

void __fastcall TfrmDemoBaisicMain::SyncElementsVisibilityWithMenu()
{
  PivotGrid()->OptionsView->ColumnFields = miShowColumnFields->Checked;
  PivotGrid()->OptionsView->DataFields = miShowDataFields->Checked;
  PivotGrid()->OptionsView->FilterFields = miShowFilterFields->Checked;
  PivotGrid()->OptionsView->FilterSeparator = miShowFilterSeparator->Checked;
  PivotGrid()->OptionsView->RowFields = miShowRowFields->Checked;
}

void __fastcall TfrmDemoBaisicMain::SyncTotalVisibilityWithMenu()
{
  PivotGrid()->OptionsView->ColumnTotals = miShowColumnTotals->Checked;
  PivotGrid()->OptionsView->RowTotals = miShowRowTotals->Checked;
  PivotGrid()->OptionsView->ColumnGrandTotals = miShowColumnGrandTotals->Checked;
  PivotGrid()->OptionsView->RowGrandTotals = miShowRowGrandTotals->Checked;
  PivotGrid()->OptionsView->TotalsForSingleValues = miShowTotalsForSingleValues->Checked;
  PivotGrid()->OptionsView->GrandTotalsForSingleValues = miShowGrandTotalsForSingleValues->Checked;
}

void __fastcall TfrmDemoBaisicMain::SyncMenuWithElementsVisibility()
{
  miShowColumnFields->Checked = PivotGrid()->OptionsView->ColumnFields;
  miShowDataFields->Checked = PivotGrid()->OptionsView->DataFields;
  miShowFilterFields->Checked = PivotGrid()->OptionsView->FilterFields;
  miShowFilterSeparator->Checked = PivotGrid()->OptionsView->FilterSeparator;
  miShowRowFields->Checked = PivotGrid()->OptionsView->RowFields;
}

void __fastcall TfrmDemoBaisicMain::SyncMenuWithTotalVisibility()
{
  miShowColumnTotals->Checked = PivotGrid()->OptionsView->ColumnTotals;
  miShowRowTotals->Checked = PivotGrid()->OptionsView->RowTotals;
  miShowColumnGrandTotals->Checked = PivotGrid()->OptionsView->ColumnGrandTotals;
  miShowRowGrandTotals->Checked = PivotGrid()->OptionsView->RowGrandTotals;
  miShowTotalsForSingleValues->Checked = PivotGrid()->OptionsView->TotalsForSingleValues;
  miShowGrandTotalsForSingleValues->Checked = PivotGrid()->OptionsView->GrandTotalsForSingleValues;
}

//---------------------------------------------------------------------------
